JSDemos/Demos/Charts/ExportCustomMarkup/React/App.js

```diff
@@ -15,77 +15,69 @@ function prepareMarkup(chartSvg, markup) {
 }
 
-class App extends React.Component {
-  constructor(props) {
-    super(props);
+function App() {
+  const [childRef, setChildRef] = React.useState(null);
+  const chartRef = React.useRef();
 
-    this.chartRef = React.createRef();
-    this.onClick = this.onClick.bind(this);
-    this.onRef = this.onRef.bind(this);
-  }
-
-  render() {
-    return (
-      <div id="chart-demo">
-        <div className="chart_environment">
-          <Form onRef={this.onRef} />
-          <Chart
-            ref={this.chartRef}
-            id="chart"
-            dataSource={dataSource}
-            palette="Violet"
-          >
-            <CommonSeriesSettings barPadding={barPadding} argumentField="state" type="bar" />
-            <Series valueField="year1990" name="1990" />
-            <Series valueField="year2000" name="2000" />
-            <Series valueField="year2010" name="2010" />
-            <Series valueField="year2016" name="2016" />
-            <Series valueField="year2017" name="2017" />
-            <Legend verticalAlignment="bottom" horizontalAlignment="center" />
-            <Title text="Oil Production">
-              <Subtitle text="(in millions tonnes)" />
-            </Title>
-          </Chart>
-        </div>
-        <div className="controls-pane">
-          <Button
-            id="export"
-            width={145}
-            icon="export"
-            text="Export"
-            type="default"
-            onClick={this.onClick}
-          />
-        </div>
-      </div>
-    );
-  }
-
-  get chart() {
-    return this.chartRef.current.instance;
-  }
-
-  onClick() {
-    exportFromMarkup(prepareMarkup(this.chart.svg(), this.child.getElement().innerHTML), {
-      width: 820,
-      height: 420,
-      margin: 0,
-      format: 'png',
-      svgToCanvas(svg, canvas) {
-        return new Promise((resolve) => {
-          toCanvas(canvas, new XMLSerializer().serializeToString(svg), {
-            ignoreDimensions: true,
-            ignoreClear: true,
-            renderCallback: resolve,
+  const onClick = React.useCallback(() => {
+    exportFromMarkup(
+      prepareMarkup(chartRef.current.instance.svg(),
+        childRef.current.getElement().innerHTML), {
+        width: 820,
+        height: 420,
+        margin: 0,
+        format: 'png',
+        svgToCanvas(svg, canvas) {
+          return new Promise((resolve) => {
+            toCanvas(canvas, new XMLSerializer().serializeToString(svg), {
+              ignoreDimensions: true,
+              ignoreClear: true,
+              renderCallback: resolve,
+            });
           });
-        });
+        },
       },
-    });
-  }
+    );
+  }, [childRef]);
+
+  const onRef = React.useCallback((ref) => {
+    setChildRef(ref);
+  }, []);
 
-  onRef(arg) {
-    this.child = arg;
-  }
+  return (
+    <div id="chart-demo">
+      <div className="chart_environment">
+        <Form onRef={onRef} />
```

Comment:
Whats this for? i wasnt able to find onRef in our docs
its a bit strange that we store ref in state and set it this way

Reply:
It is custom form component

Reply:
ok, found out that is a prop for custom form component
maybe do the same with forwardRef?